Added a SIGSEGV handler that dumps the stacktrace to ease reporting

marcelofg55 opened the pull request:
This PR adds a simple crash reporter that dumps the stacktrace to stderr, implemented for various OSes:

- On Linux it adds a signal for SIGSEGV/SIGFPE that prints the stacktrace when the signal is triggered. It uses addr2line to get file/line number.

- On OS X it also adds a signal for SIGSEGV/SIGFPE that prints the stacktrace when the signal is triggered. It uses atos to get file/line number.

- On Windows it uses __try/__except to catch exceptions. Only supported on MSVC (no mingw support yet).

Note, the code is only enabled with crash_report=yes on debug mode, also after the stacktrace is generated the crash is passed to the OS so that you can debug if necessary.

This is a screenshot of how it shows the stacktrace on Linux:
![screenshot from 2017-08-06 17-18-12](https://user-images.githubusercontent.com/10578225/29006848-30865dfa-7ace-11e7-9586-e4aaf547f1ca.png)

On Windows:
![crashreport](https://user-images.githubusercontent.com/10578225/29011895-37c4df9e-7b0d-11e7-81e7-b2d5f44960dd.png)

On OS X:
![screen shot 2017-08-07 at 1 36 55 pm](https://user-images.githubusercontent.com/10578225/29036789-43807bd0-7b77-11e7-8a5f-4bb83e50d306.png)

marcelofg55 committed:
Added a crash handler to dump the backtrace on Windows, Linux and OS X

reduz commented:
I'm not sure about it, generally Linux users know how obtain a backtrace
and traces are, in most cases, not really as useful as conditions to
reproduce it

On Sun, Aug 6, 2017 at 11:35 AM, Marcelo Fernandez <notifications@github.com
> wrote:

> This PR adds a signal handler for SIGSEGV/SIGFPE that prints the
> stacktrace when the signal is triggered, it works on Linux/OS X (I'm
> working on a similar code for Windows).
>
> Note, the code is only enabled on debug mode.
>
> This is a screenshot of how it shows the stacktrace:
> [image: screen shot 2017-08-06 at 11 22 17 am]
> <https://user-images.githubusercontent.com/10578225/29004163-9c4dc242-7a99-11e7-9c4c-696457d9a06e.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/godotengine/godot/pull/10124
> Commit Summary
>
>    - Added a SIGSEGV handler that dumps the stacktrace to ease reporting
>
> File Changes
>
>    - *M* drivers/unix/os_unix.cpp
>    <https://github.com/godotengine/godot/pull/10124/files#diff-0> (22)
>    - *M* platform/osx/detect.py
>    <https://github.com/godotengine/godot/pull/10124/files#diff-1> (1)
>    - *M* platform/x11/detect.py
>    <https://github.com/godotengine/godot/pull/10124/files#diff-2> (1)
>
> Patch Links:
>
>    - https://github.com/godotengine/godot/pull/10124.patch
>    - https://github.com/godotengine/godot/pull/10124.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/10124>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z29oxF37XCtClVwpA--vGIc4Wnxc8ks5sVc8ygaJpZM4Ouv1e>
> .
>

marcelofg55 commented:
Yeah, usually Linux users do know how to use gdb and such but I've seen many crash Issues reported without a stacktrace, so I thought this could help users to easily report a crash with the stacktrace. I agree that conditions to reproduce are usually more helpful to solve the crashes, but with a stacktrace included with the Issue report it lets you know where the problem might be.

reduz commented:
how do crash reporters of other apps work? also, i guess getting the
argument values and real lines of code is impossible right? as is those
traces don't seem too useful

On Sun, Aug 6, 2017 at 12:15 PM, Marcelo Fernandez <notifications@github.com
> wrote:

> Yeah, usually Linux users do know how to use gdb and such but I've seen
> many crash Issues reported without a stacktrace, so I thought this could
> help users to easily report a crash with the stacktrace. I agree that
> conditions to reproduce are usually more helpful to solve the crashes, but
> with a stacktrace included with the Issue report it lets you know where the
> problem might be.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/10124#issuecomment-320513067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z25BkNwyRSXckzGx2R6JjCPQBhUYEks5sVdh4gaJpZM4Ouv1e>
> .
>

Zireael07 commented:
I agree that it's easier to set up gdb on Linux, but a crash reporter of some kind for Windows would be lovely (aaargh gdb and Windows aaargh)

marcelofg55 commented:
Lines of code seems possible on Linux, need to work it out.

marcelofg55 commented:
I was just looking into how other apps handle crashes and for example I've found that unreal has this type of crash reporter, seems pretty cool in my opinion: https://answers.unrealengine.com/storage/attachments/41297-ue4_crash.png

marcelofg55 commented:
Update: I managed to improve the stacktrace output on Linux, now the function name is demangled and the file/line number is shown as well:

![screenshot from 2017-08-06 17-18-12](https://user-images.githubusercontent.com/10578225/29006690-5252a34c-7acb-11e7-8967-8bc72303e014.png)

bruvzg commented:
@marcelofg55 `atos -o execpath addr` should give line number on macOS in the same manner as `addr2line` do on Linux

marcelofg55 commented:
@bruvzg thanks for the tip, I'll look into it

neikeq commented:
@reduz Godot binaries built with mono also have a SIGSEGV handler from mono that dumps the stacktrace, this is an example report: https://gist.github.com/neikeq/5002fd23f9d6d2f3843257d9326bcb42

marcelofg55 commented:
Another update, been working on the windows code for this, here's how the crash report is looking so far:
![crashreport](https://user-images.githubusercontent.com/10578225/29011640-12ca1ba2-7b0b-11e7-9360-2525ca308ef4.png)

akien-mga commented:
This is awesome! :)

If you need references for cross-platform (especially Windows) stack traces, you can check:
- https://github.com/OpenDungeons/OpenDungeons/blob/development/source/utils/StackTraceUnix.cpp
- https://github.com/OpenDungeons/OpenDungeons/blob/development/source/utils/StackTraceWinMSVC.cpp
- https://github.com/OpenDungeons/OpenDungeons/blob/development/source/utils/StackTraceWinMinGW.cpp

The code is GPL, but I know the author and I'm positive he'd agree to have us draw ideas from it/reuse some chunks under the MIT, if needed. (Haven't checked your code yet, maybe you already have a better feature ;)).

marcelofg55 commented:
Thanks for the references @akien-mga! The current implementation on this PR lacks mingw support yet so I'm going to check how the Open Dungeons code works for that one, and I'll check the others for ideas as well :).

akien-mga commented:
> The code is GPL, but I know the author and I'm positive he'd agree to have us draw ideas from it/reuse some chunks under the MIT, if needed. (Haven't checked your code yet, maybe you already have a better feature ;)).

CC'ing the author @hwoarangmy (hi! how are you doing? :D) for confirmation that we can use his GPL'ed code as reference for a MIT implementation.

marcelofg55 commented:
Yet another update, this time with the feature for OS X (it uses atos to get file/line number as suggested by @bruvzg): 
![screen shot 2017-08-07 at 1 36 55 pm](https://user-images.githubusercontent.com/10578225/29036353-80dade64-7b75-11e7-9a4e-3bc7962160cd.png)

reduz commented:
Some people seems to find this interesting, despite that it has many drawbacks. Given it's pretty simple, I suggest you add a compile option for it and leave it off by default.. then let people experiment with it.

neikeq commented:
What are the drawbacks?

marcelofg55 commented:
@reduz Alright, I've changed the commit so that this is only enabled when you use crash_report=yes with scons (defaults to no).

karroffel commented:
This doesn't seem to handle ARM processors

marcelofg55 commented:
@karroffel ARM processors on what OS you mean?

karroffel commented:
OSX it seems https://github.com/godotengine/godot/pull/10124/files#diff-9d98950bd0f7189e0e06382464feb2e2R120

marcelofg55 commented:
Are there Macs that come with ARM processor?

karroffel commented:
Ups, never mind 😄 🙈

marcelofg55 commented:
Haha :)

marcelofg55 commented:
I've tested this feature while trying to fix #10157, to see if this was useful of real life crashes, and it helped me to fix it without the need of using gdb :)

Here's the screenshot of the reported crash:
![screen shot 2017-08-08 at 5 50 14 pm](https://user-images.githubusercontent.com/10578225/29094669-81c79f02-7c64-11e7-8ecd-7d71877b1dc5.png)

tagcup commented:
I really like this feature but I think it's better to remove the "please include this when reporting" part considering this might be shipped with actual games, and they may have different bug report policies.

marcelofg55 commented:
@tagcup Actually this is only enabled with target=debug (and crash_report=yes), so I don't think it will be shipped with games, but I can remove that part of the message anyways.

tagcup commented:
@marcelofg55 Yes, I knew that. They can choose to enable backtrace, but backtrace on sigsegv and crash reports to developers are orthogonal things.

Also, I think the option could be renamed to backtrace too.

marcelofg55 commented:
Sure, I've just changed the message.

akien-mga commented:
> I really like this feature but I think it's better to remove the "please include this when reporting" part considering this might be shipped with actual games, and they may have different bug report policies.

Well, if the backtrace is shown in exported games, I would expect users to include it when reporting it to the game developer. The developer can then decide if it's meaningful or not to make an upstream bug report - to the gamer, it should not matter that it's a crash of Godot, but of their game.

So the proposed message was fine IMO. There could be a project option to define a bug report URL / email address, so that you can make the message something like:

- by default: "please include this when reporting the bug on https://github.com/godotengine/godot/issues"
- in an exported game: "please include this when reporting the bug to support@myawesomegame.com"

akien-mga commented:
The main concern @reduz had about this feature is that if you catch SIGSEGV, it might make it cumbersome to really debug issues with gdb. Can you confirm that this would be an issue, or would gdb have the priority and catch the segfault before your logic gets triggered?

I personally think this feature would be nice to have by default on release builds (both for the editor and the template), so if the above concern is valid, I would propose this logic:
- add a command line switch (in main.cpp) to enable/disable the feature
- in `tools=yes target=release_debug` and `tools=no target=release` (release builds of the editor and templates, i.e. the binaries we distribute), enable the feature by default - it can be disabled via the command line switch if needed
   - before doing that, it would be good to assess how useful the backtraces are in such binaries though, as we strip some of the debug symbols
- in `tools=yes target=debug` and `tools=no target=release_debug` (debug builds of the editor and templates), disable the feature by default so that they can be debugged via gdb/windbg - it can be enabled via the command line switch if needed

marcelofg55 commented:
@akien-mga @reduz I've just tested on OS X and Linux, and on both cases gdb has priority over the handle_sigsegv code, so the function doesn't get in the way. So I guess that the command line switch is not needed then?

marcelofg55 commented:
As suggested by @akien-mga I've added another option to scons for the crash message using crash_report_message="message here" (defaults to "Please include this when reporting the bug on https://github.com/godotengine/godot/issues")

akien-mga commented:
> As suggested by @akien-mga I've added another option to scons for the crash message using crash_report_message="message here" (defaults to "Please include this when reporting the bug on https://github.com/godotengine/godot/issues")

I wanted this as a project-specific option actually (in ProjectSettings), so that users don't have to recompile their export templates just for that. If that's not possible (e.g. ProjectSettings not initialized at that stage), it's probably better to drop the option altogether and use a generic message. Our compiler defines are way too long already :)

akien-mga commented:
So if it does not conflict with normal debugging, I would just make this enabled by default (without scons option to disable it). Then I'd add an option to the Godot binary (via main.cpp) to disable it when needed (e.g. `-disable-backtrace`).

marcelofg55 commented:
So I've added that -disable-backtrace and the project setting for the message as well. And removed the scons previous stuff.
By the way, this is tagged as 3.0 but I've made the PR for the 2.1 branch (I will port these changes for the master branch after this gets approved of course).

hpvb commented:
Sorry, this is all fine for 2.1. Are you planning to do a PR for 3.0 too?

hpvb commented:
There are some concerns about adding all of these to the platform's main files.

Would you mind splitting out the implementations in each platform/ directory? Maybe have a platform/name/crash_handler.cpp/h and just call it from main?

Thanks! I really love this feature though.

hpvb commented:
Would you mind splitting the crash handler code in the 2.1 version also?

marcelofg55 commented:
@hpvb Just changed the commit with the crash handler .cpp/h files.

hpvb commented:
I think we're still discussing what we want to do for 2.1.

akien-mga commented:
Are some fixes needed based on early feedback in the master branch? (some compiles fixes + preventing blocking the debugger on MSVC)
I'll merge once those are in.

marcelofg55 commented:
Just added those changes.

akien-mga commented:
Thanks, I'll merge once Travis has finished building.
